### ISSUE TITLE
profile guided optimization: option to disable inserting stack frames (popFrame) in hot spots only

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/Nim#8959]
TLDR:
* `--stacktrace:on` can be very expensive (6X slowdown in simple real example below), due to inserted `popFrame` being expensive
* for top performance / benchmark comparisons, we should mention in docs to run with `-d:release --stacktrace:off`, not just `-d:release` (even if `--stacktrace:off` is default, it may be in user's nim.cfg); often I only see `-d:release` mentioned in nim forums, etc
* using`--stacktrace:off` may be too blunt, see proposal below for another approach

## motivation
Consider the following code [1], which computes an integral image and reports running time over a number of iterations.
* with `nim c -r -d:case1 -d:release main.nim`, performance is on par with equivalent C++ code.
* `--stacktrace:on` or --stacktrace:off` makes no difference

Now consider the same example with -d:case2 instead of -d:case1 (ie, using float32 instead of int32)
* with `--stacktrace:off`, performance is on par with C++
* with `--stacktrace:on`, performance is 7.3 seconds instead of 1.13 seconds, a **6.4X slowdown**!

Upon investigation, the culprit is `+=` which, for float (but not int), gets codegened in C to a function `pluseq__XXX` instead of being inlined to `+=`, see [2].
This is in turn due to how += is defined in stdlib, being differently handled for int vs float, see [3]
I'm actually not sure why float can't be handled the same way as it is for int (via a magic), it could improve performance for apps built with `--stacktrace:on`.

But that's just 1 example and there's always gonna be places where a codegen'd function is needed anyways. On the other hand disabling stacktraces altogether may be undesirable, so instead my proposal below.

## proposal (finally...): use profile guided optimization
introduce a new flag `--pgo:[OPTIONS...]` that selectively removes the inserted `nimfr_` / `popFrame` stack frame logic in hot-spots only (ie, in a specific code location), using user provided profile file (from a previous run) and threshold. Other optimizations could be done in the future as well, but stack frame insertion can have a major performance impact so we can start with that.

## [1] the code
```nim
# D20180913T001057
import times

proc integralImage5[T](data: seq[T], data2: var seq[T], p, q: int)=
  var temp: T
  var rowsums = newSeq[T](p)
  var k=0
  for j in 0..<q:
    temp = 0
    for i in 0..<p:
      rowsums[i]+=data[k]
      temp+=rowsums[i]
      data2[k]=temp
      k.inc

proc main=
  block:
    when defined(case1):
      type T=int32
    when defined(case2):
      type T=float32
    let p=1000
    let q=p

    let numIter=1000
    let n = p*q
    var data = newSeq[T](n)
    for i in 0..<n:
      data[i]=T(i mod 10)
    var data2 = newSeq[T](n)
    var rowsums = newSeq[T](p)

    let t0 = epochTime()
    for i in 0..<numIter:
      integralImage5(data, data2, p, q)
    let elapsed = epochTime() - t0

    var total=0.T
    # prevent optimizing out computation
    for i in 0..<n:
      total+=data2[i]

    echo (elapsed:elapsed, total:total)

main()
```

## [2] C codegen for += for float (case2)

* with --stacktrace:off
```c
static N_INLINE(void, pluseq__0Q9c9ccOr82GxaUx3qtv9aCoQmain2)(NF32* x, NF32 y) {
  (*x) = ((NF32)((*x)) + (NF32)(y));
}
```

* with --stacktrace:on
```c
static N_INLINE(void, pluseq__0Q9c9ccOr82GxaUx3qtv9aCoQmain2)(NF32* x, NF32 y) {
  nimfr_("+=", "/Users/timothee/git_clone/nim/Nim/lib/system.nim");
  nimln_(3700, "/Users/timothee/git_clone/nim/Nim/lib/system.nim");
  (*x) = ((NF32)((*x)) + (NF32)(y));
  popFrame();
}
```
## [3] += in stdlib differs for int vs float
```nim
proc `+=`*[T: SomeOrdinal|uint|uint64](x: var T, y: T) {. magic: "Inc", noSideEffect.}
proc `+=`*[T: float|float32|float64] (x: var T, y: T) {. inline, noSideEffect.} = x = x + y
```
+= is magic for int only; for float it's a regular proc.

**Comment by Araq** [comment]
What's wrong with ``{.push stackTrace: off.}`` ?

**Comment by timotheecour** [comment]
`{.push stackTrace: off.} ` is a good workaround when we know ahead of time the performance bottlenecks; but oftentimes it's best to measure (via actual profiling) as guessing is usually misleading, in particular performance bottlenecks change depending on how a program is being used for a particular application.
Also, using PGO avoids having to litter source code with performance related pragmas such as `{.push stackTrace: off.}`

**Comment by Araq** [comment]
I don't understand what you're proposing here.

> for top performance / benchmark comparisons, we should mention in docs to run with -d:release --stacktrace:off, not just -d:release (even if --stacktrace:off is default, it may be in user's nim.cfg); 

Yeah, but so what. Maybe don't put bad things into your ``nim.cfg`` then.

**Comment by awr1** [comment]
I'm not too familiar with using PGO in practice, but I think if PGO options ever get added to the compiler we might as well throw in invocation of the C compiler's PGO systems.
(e.g. `-fprofile-XXX` on gcc and `-GEN/USEPROFILE` on msvc)

**Comment by timotheecour** [comment]
> I don't understand what you're proposing here.

I'm proposing an enhancement that will allow generating efficient code even in the presence of `--stacktrace:on`, without having to litter the source code with `{.push stackTrace: off.}` on all the places that could be bottlenecks.
When a program crashes with no stacktrace (eg in customer's hands), it can be hard to trace back where the problem was; but, as I show above, `--stacktrace:on` can incur a large performance hit, leaving the choice between:
* A: finding all the potential hotspots and annotating the code "by hand" with `{.push stackTrace: off.}` in hotspots
* B: automating A without having to edit source code, using this proposal

I have not given details on how B could be implemented, happy to discuss that more, but in short, it would allow user to specify a 
```nim
nim c --profile:on --stacktrace:on main.nim arg1 arg2 # modulo fixing https://github.com/nim-lang/Nim/issues/8991
# this would generate profile_results.txt

nim c --pgo:profile:profile_results.txt,threshold:0.1 --stacktrace:pgo main.nim
```
this would avoid emitting the stubs `nimfr_` / `popFrame` for hot-spot functions, as specified by user supplied threshold 0.1 (ie functions taking up 10% of total runtime)

**Comment by awr1** [comment]
What I was referring to was PGO as a general concept (e.g. branch prediction hints, as you said) and not the stack frame optimization specifically.
